File: Skyscraper-iOS-20251106/Skyscraper/Skyscraper/Models/PostDraft.swift
//
//  PostDraft.swift
//  Skyscraper
//
//  Model for saving draft posts
//

import Foundation
import Combine

struct PostDraft: Codable, Identifiable {
    let id: UUID
    var text: String
    var imageData: [Data] // Store image data as Data arrays
    var imageAltTexts: [String]
    var createdAt: Date
    var updatedAt: Date
    var languageId: String
    var moderationSettings: PostModerationSettings

    init(
        id: UUID = UUID(),
        text: String,
        imageData: [Data] = [],
        imageAltTexts: [String] = [],
        languageId: String,
        moderationSettings: PostModerationSettings
    ) {
        self.id = id
        self.text = text
        self.imageData = imageData
        self.imageAltTexts = imageAltTexts
        self.createdAt = Date()
        self.updatedAt = Date()
        self.languageId = languageId
        self.moderationSettings = moderationSettings
    }

    var preview: String {
        if text.isEmpty {
            return "Empty draft"
        }
        return text.count > 100 ? String(text.prefix(100)) + "..." : text
    }

    var relativeTime: String {
        let interval = Date().timeIntervalSince(updatedAt)
        let minutes = Int(interval / 60)
        let hours = Int(interval / 3600)
        let days = Int(interval / 86400)

        if days > 0 {
            return "\(days)d ago"
        } else if hours > 0 {
            return "\(hours)h ago"
        } else if minutes > 0 {
            return "\(minutes)m ago"
        } else {
            return "Just now"
        }
    }
}

// MARK: - Draft Manager
@MainActor
class DraftManager: ObservableObject {
    static let shared = DraftManager()

    @Published var drafts: [PostDraft] = []

    private let userDefaultsKey = "com.skyscraper.postDrafts"
    private let maxDrafts = 50

    private init() {
        loadDrafts()
    }

    func saveDraft(_ draft: PostDraft) {
        // Update existing draft or add new one
        if let index = drafts.firstIndex(where: { $0.id == draft.id }) {
            var updatedDraft = draft
            updatedDraft.updatedAt = Date()
            drafts[index] = updatedDraft
        } else {
            drafts.insert(draft, at: 0)
        }

        // Keep only the most recent drafts
        if drafts.count > maxDrafts {
            drafts = Array(drafts.prefix(maxDrafts))
        }

        persistDrafts()
    }

    func deleteDraft(_ draft: PostDraft) {
        drafts.removeAll { $0.id == draft.id }
        persistDrafts()
    }

    func getDraft(id: UUID) -> PostDraft? {
        drafts.first { $0.id == id }
    }

    private func loadDrafts() {
        guard let data = UserDefaults.standard.data(forKey: userDefaultsKey),
              let decoded = try? JSONDecoder().decode([PostDraft].self, from: data) else {
            drafts = []
            return
        }
        drafts = decoded
    }

    private func persistDrafts() {
        if let encoded = try? JSONEncoder().encode(drafts) {
            UserDefaults.standard.set(encoded, forKey: userDefaultsKey)
        }
    }
}

File: app/src/main/java/com/cameronbanga/skyscraper/models/PostDraft.kt
package com.cameronbanga.skyscraper.models

import com.squareup.moshi.JsonClass
import java.time.Instant
import java.util.UUID

@JsonClass(generateAdapter = true)
data class PostDraft(
    val id: String = UUID.randomUUID().toString(),
    var text: String,
    var imageData: List<ByteArray> = emptyList(),
    var imageAltTexts: List<String> = emptyList(),
    val createdAt: Long = System.currentTimeMillis(),
    var updatedAt: Long = System.currentTimeMillis(),
    var languageId: String,
    var moderationSettings: PostModerationSettings
) {
    val preview: String
        get() = if (text.isEmpty()) {
            "Empty draft"
        } else {
            if (text.length > 100) text.take(100) + "..." else text
        }

    val relativeTime: String
        get() {
            val now = System.currentTimeMillis()
            val interval = (now - updatedAt) / 1000  // Convert to seconds
            val minutes = interval / 60
            val hours = interval / 3600
            val days = interval / 86400

            return when {
                days > 0 -> "${days}d ago"
                hours > 0 -> "${hours}h ago"
                minutes > 0 -> "${minutes}m ago"
                else -> "Just now"
            }
        }
}
